src/memory-fs.js

```diff
@@ -0,0 +1,199 @@
+const path = require('path');
+
+const { createHash } = require(`crypto`);
+const errors = require('errno');
+const util = require('util');
+const { ReadableStream, WritableStream } = require('stream');
+const MemoryFileSystemError = require('memory-fs/lib/MemoryFileSystem');
+
+const debug = util.debuglog('astroturf:memory-fs');
+
+const returnsTrue = () => true;
+const returnsFalse = () => false;
+
+const md5 = input => {
+  const hash = createHash('md5');
+  hash.update(input);
+  return hash.digest('hex');
+};
+
+const read = (file, optsOrEncoding) => {
+  const encoding = optsOrEncoding?.encoding || optsOrEncoding;
+
+  file.atime = new Date();
+
+  return encoding ? file.contents.toString(encoding) : file.contents;
+};
+
+/**
+ * A simple in memorry ponyfill of the node fs that stores objects in memory.
+ * We don't use `memory-fs` because our use-case is much narrower (only need files not dirs)
+ * and it doesn't support stat timestamps
+ */
+class MemoryFs {
+  constructor() {
+    this.paths = new Map();
+
+    ['stat', 'readdir', 'rmdir', 'unlink', 'readFile', 'writeFile'].forEach(
+      fn => {
+        this[fn] = (...args) => {
+          const cb = args.pop();
+          let result;
+          try {
+            result = this[`${fn}Sync`](...args);
+          } catch (err) {
+            setImmediate(() => cb(err));
+            return;
+          }
+          setImmediate(() => cb(null, result));
+        };
+      },
+    );
+  }
+
+  addFile = (p, data, updateMtime = false) => {
+    p = path.normalize(p);
+    const hash = md5(data);
+    const existing = this.paths.get(p);
+
+    const keepTime = !updateMtime && existing && existing.hash === hash;
+    if (!keepTime)
+      debug(
+        `${existing ? 'modifying' : 'writing'} file ${path.relative(
+          process.cwd(),
+          p,
+        )} [${hash}]`,
+      );
+
+    this.paths.set(p, {
+      hash,
+      contents: Buffer.from(data),
```

Comment:
is there an explicit reason to create the buffer eagerly, rather than just keeping around `data` and creating the mutable buffer only as needed?

Reply:
no reason, that's a good idea